[#9] Bugfix: Patch up passing strings with unescaped quotes on outputs

bronius opened the pull request:
** UPDATE: 2025-10-30** This PR introduces all the breaking. Not ready for review. Conceptual guidance still welcome in the interim, [per comment requesting it](https://github.com/tamu-edu/it-ae-actions-terraform-pr-apply/pull/10#issuecomment-3463377655).

# Description

This PR fixes an issue where special characters in Terraform outputs (specifically _parentheses_ in particular) were causing syntax errors in the GitHub Actions runner's generated shell scripts. The generated github runner shell script had lines like:
```
echo "....anything goes here (except this) which would fail"
```
which would trip it up. Even quotes (`"`) and all types of other questionable characters were fine -- _only parentheses_-- with an error like:
```
Run touch apply.log
/home/runner/work/_temp/937b6b77-1cdc-44d2-91a8-de4eb8618637.sh: line 164: syntax error near unexpected token `('
```
in the calling workflow.

Solution:

* Swapped out some steps of this composite action for off-the-shelf module `op5dev/tf-via-pr@v13` (see https://github.com/OP5dev/TF-via-PR).
* Preserved visibility by still displaying output in the console while running (although now it comes out all at once)
* Preserved ability to pass in `-var-file` and `-backend-config`

See:
https://github.com/search?q=org%3Atamu-edu+terraform-init-flags&type=code
https://github.com/search?q=org%3Atamu-edu+terraform-apply-flags&type=code
for justification of how `arg-var-file` and `arg-backend-config` were implemented.

This has not been tested again since the re-addition of s3 upload, nor has the specific, newly added functionality been tested at all.

## How to Test

Here is a run that did not work, specifically because of the following line (and a similar one just below it, but it was never reached), but now does:
```
"name" = "ITAS - ReviewEm test - Web Frontend (Client ID)"
```
https://github.com/tamu-edu/it-eas-reviewem-iac/actions/runs/18829086771/job/53717195002#step:5:1442

But when invoked using the branch on this PR it succeeds:
https://github.com/tamu-edu/it-eas-reviewem-iac/actions/runs/18890215217
(and then some! It now comments outcome back on the merged PR (see https://github.com/tamu-edu/it-eas-reviewem-iac/pull/193#issuecomment-3458916768)!)

Resolves #9.

bronius committed:
[#9] Bugfix: Patch up passing strings with unescaped quotes on outputs

bronius committed:
Fixes heredoc fix implementation

bronius committed:
Trying quotes instead of apostrophes around grep in summary

bronius committed:
Heavier use of HEREDOC to try to avoid interruptive characters

bronius committed:
Yaml didn't like that format of heredocs. Trying a nuclear sed instead.

bronius committed:
Trying base64 encode to avoid funny characters interfering with generated github runner shell script

bronius committed:
Replace convoluted and dicy tee approach for straightforward temp files solution

bronius committed:
Introducing intermediary variables, bc suspecting overwriting stdout/stderr is bad news

bronius commented:
@joraff not sure how to add you as a reviewer here

joraff commented on action.yml:
```diff
@@ -92,8 +92,30 @@ runs:
+        stderr_file=$(mktemp)
+
+        # Run terraform with outputs redirected to files
+        terraform apply -input=false -no-color ${{ inputs.auto-approve == 'true' && '-auto-approve' || '' }} ${{ inputs.terraform-apply-flags }} > "$stdout_file" 2> "$stderr_file"
```
This can be done better with tee, which will both stream to stdout and write to a file at the same time. My concern with below is having to wait until the process is complete to see the output. For long applies or refreshes, I think lots of time and worry will be spent that terraform didn't actually run (or is hung)

joraff replied:
You can also skip the passing around the outputs altogether, too, if the outputs are already written to file in the working directory.

bronius replied:
I agree that it would be better to have realtime logging feedback: You can see [in an earlier commit](https://github.com/tamu-edu/it-ae-actions-terraform-pr-apply/pull/10/commits/aaebdf86e5e05c2417f186adbd4da43323764a08) that I tried `tee` in at least one of my first 100 stabs but ran into some silly trail mix of dual output and base64 encoding and plaintext which just did not end up right. I'll give it another go.

bronius committed:
Adds pseudo unit tests to help TDD

bronius committed:
Git to show in my fork Actions

bronius committed:
I just want to demonstrate that it fails one time and see how it looks at Github. Then I'll change it to a failing test.

bronius committed:
[#9] Fix parenthesis breaks action by swapping out method for op5dev/tf-via-pr@v13

bronius committed:
Hallucination fix 1: Map new module's parameters better (WIP)

bronius committed:
Attempt to fix apply to run plan+apply in a single step

bronius committed:
Attempt to fix apply to run plan+apply in a single step (WIP)

bronius committed:
Updates to more fitting unit-tests

bronius committed:
Remove failing but unrelated failing Verification step

bronius committed:
Switch unit test to pass values by file rather than fragile string redirects

bronius committed:
Remove automatic run of unit test. It's more for demonstration and not so much stability.

bronius commented:
@joraff Please see this draft PR and weigh in on:
* ~Should I put the s3 upload back in? Copilot did not itself preserve the functionality, and I did a quick scan [across Tamu projects](https://github.com/search?q=org%3Atamu-edu%20it-ae-actions-terraform-pr-apply&type=code) and don't see any implementation of it (in apply workflows). I do see it filling a purpose (to preserve a plan after having applied),  but I wonder if _we need_ it.~
I went ahead and added it back in (accf9e735a04bd44380c1e1dd8ff52c8733edf8d) from https://github.com/tamu-edu/it-ae-actions-terraform-pr-apply/blob/main/action.yml#L116-L150. I tried and tried to preserve your original authorship on those lines, but no dice. The primary difference was to adapt it to this module with `${{ inputs.working-directory }}/apply.log` as a source in that step.
* Also copilots doing: ~We have orphaned a few parameters like tf init arguments, and working-directory now serves as both working directory for running cli cmds and for running terraform. Seems to be an improvement, but maybe you can think of use case(s) where we would want to keep both arguments.~
I've handled this. See latest commits for how.

~Actually.. I see now that `terraform-workspace` is the "dev, test, prod" for `terraform workspace`. Our latest model happens to always used `default` workspace within an environment, so the separation should still be preserved (and properly implemented with this module).~
I have added this back, too.

See my updated PR description for add'l details to call out and examples demonstrating this updated composite action works.

bronius committed:
Adds support for terraform-backend-config and brings back deprecated terraform-init-flags to support it

bronius committed:
Adds support for terraform-var-file and brings back deprecated terraform-apply-flags to support it

bronius committed:
Bring back s3 upload of apply.log (Sorry, Joseph, I tried and tried to preserve original authorship blame but failed)

bronius committed:
Possible fix to use of unsupported split() function

bronius committed:
Revert "Possible fix to use of unsupported split() function"

This reverts commit 34524008fcc6efebfc23d3e180b901bf397a90b5.

bronius committed:
Revert "Bring back s3 upload of apply.log (Sorry, Joseph, I tried and tried to preserve original authorship blame but failed)"

This reverts commit accf9e735a04bd44380c1e1dd8ff52c8733edf8d.

joraff commented on action.yml:
```diff
@@ -93,46 +95,6 @@ runs:
-      with:
-        terraform_version: ${{ inputs.terraform-version }}
-
-    - name: Parse legacy terraform flags
-      shell: bash
-      id: parse_flags
-      run: |
-        # Parse backend config from deprecated terraform-init-flags
-        backend_config=""
-        if [[ -n "${{ inputs.terraform-backend-config }}" ]]; then
-          backend_config="${{ inputs.terraform-backend-config }}"
-        elif [[ -n "${{ inputs.terraform-init-flags }}" ]] && [[ "${{ inputs.terraform-init-flags }}" == *"-backend-config="* ]]; then
-          # Extract the backend config value from the flags
-          init_flags="${{ inputs.terraform-init-flags }}"
-          # Find the backend-config parameter and extract its value
-          if [[ $init_flags =~ -backend-config=([^[:space:]]+) ]]; then
-            backend_config="${BASH_REMATCH[1]}"
-          fi
-        fi
-
-        # Parse var file from deprecated terraform-apply-flags
-        var_file=""
-        if [[ -n "${{ inputs.terraform-var-file }}" ]]; then
-          var_file="${{ inputs.terraform-var-file }}"
-        elif [[ -n "${{ inputs.terraform-apply-flags }}" ]] && [[ "${{ inputs.terraform-apply-flags }}" == *"-var-file="* ]]; then
-          # Extract the var file value from the flags
-          apply_flags="${{ inputs.terraform-apply-flags }}"
-          # Find the var-file parameter and extract its value
-          if [[ $apply_flags =~ -var-file=([^[:space:]]+) ]]; then
-            var_file="${BASH_REMATCH[1]}"
-          fi
-        fi
-
-        echo "backend_config=$backend_config" >> $GITHUB_OUTPUT
-        echo "var_file=$var_file" >> $GITHUB_OUTPUT
-        echo "Parsed backend config: $backend_config"
-        echo "Parsed var file: $var_file"
-
```
I believe this was added to keep the inputs to the composite action backwards compatible while using the new terraform parser action. 

It looks like something similar is being done on L108-109, yes?

bronius replied:
That's right. See PR description for how I came to this idea: I first looked for instances of folks on campus using this module then narrowed to who's implemented these two inputs I considered for deprecation, and then I sought to implement backwards-compatibility with how they are using it. I did not test them in this way nor the intended way with the newly introduced replacement inputs.

joraff replied:
I'm okay with changing the inputs format and breaking backwards compatibility so long as we use an appropriate release version. But if it's not hard to maintain that, let's try. I guess next steps is to create some test cases to test each input?

joraff commented on action.yml:
```diff
@@ -158,50 +120,15 @@ runs:
         expand-diff: false
         expand-summary: false
 
-    - name: Generate s3 key name
-      id: s3
-      if: ${{ steps.tf-via-pr.outputs.plan-url != '' }}
-      shell: bash
-      working-directory: ${{ inputs.working-directory }}
-      run: |
-        # Download hcl2json
-        curl -Lo hcl2json $(curl -s https://api.github.com/repos/tmccombs/hcl2json/releases | jq -r '.[] | select(.tag_name == "v0.5.0").assets[] | select(.name == "hcl2json_linux_amd64").browser_download_url')
-        chmod +x hcl2json
-
-        # Get s3 bucket and key by converting terraform backend config to json
-        json=$(./hcl2json *.tf)
-
-        [[ -z "${{ inputs.s3-bucket }}" ]] \
-          && s3_bucket=$(echo $json | jq -r '.terraform[0].backend.s3[0].bucket') \
-          || s3_bucket=${{ inputs.s3-bucket }}
-
-        [[ -z "${{ inputs.s3-key }}" ]] \
-          && s3_key=$(echo $json | jq -r '.terraform[0].backend.s3[0].key') \
-          || s3_key=${{ inputs.s3-key }}
-
-        echo "S3 bucket: $s3_bucket"
-        echo "S3 key: $s3_key"
-
-        # Generate s3 key name
-        key="apply_logs/${{ github.repository }}/pr-${{ steps.pr_id.outputs.number }}/$(date -u +'%Y-%m-%d/apply-%Y-%m-%dT%H:%M:%SZ.log')"
-        echo "S3 key name: $key"
-
-        # Copy terraform apply log to s3
-        s3_url="s3://${s3_bucket}/${s3_key}/${key}"
-        aws s3 cp ${{ inputs.working-directory }}/apply.log $s3_url
-
-        echo "s3_bucket=$s3_bucket" >> $GITHUB_OUTPUT
-        echo "s3_key=$s3_key/$key" >> $GITHUB_OUTPUT
-        echo "s3_url=$s3_url" >> $GITHUB_OUTPUT
-
```
Good call to remove it, the vision for it wasn't finished. It was to store the artifacts of the terraform output in the same backend location where the state file is, since the GHA output logs do eventually expire.

Obviously not everyone will be using S3 as the backend, so it should support all our common backend locations through autodetection or inputs. So perhaps we can fork to a feature branch to keep this around for a later addition.

bronius replied:
> It was to store the artifacts of the terraform output in the same backend location where the state file is, since the GHA output logs do eventually expire.

Yep, gathered that: I saw the only implementation of it was an aws project, and it was using the `-cloud-` named fork of this module. The tf-via-pr module I implemented here can be asked for the apply log, generated plan, etc as outputs, and we could provide a singular mechanism to toss it up to s3/blob/etc in this same composite action, but I propose it be an intentional effort, well thought-out, and added when needed.
